### PR TITLE
feat(udf): Java UDF SDK

### PR DESCRIPTION
**Pull request by wangrunji0408**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

This PR adds an initial support for Java UDF. #9981

All existing tests for Python UDF are passed in Java UDF. 
The SDK API generally follows [Flink](https://nightlies.apache.org/flink/flink-docs-master/docs/dev/table/functions/udfs/#overview). See `UdfExample.java` for examples.

This PR introduced **breaking changes** to the table function API and the protocol between kernel and UDF server. cc @xxchan 
1. The output chunk now has exactly 2 columns. If the function returns multiple columns, they will be put into the second column of the chunk as a struct.
2. The datatype of the first column of the chunk is changed from int64 to int32. Because the indexes are always less than the chunk size, which is usually 1024 and will never overflow i32. Perhaps int16 (32768) is enough? 🤔

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [x] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- SQL commands, functions, and operators

### Release note

I'll draft a detailed developer guide for Java UDF in a subsequent PR.

</details>

### CODE REVIEW

**Comment by BugenZhao** [comment]
> 1. If the function returns multiple columns, they will be put into the second column of the chunk as a struct.

May I ask the reason for reverting this (#9635)? 👀

**Comment by BugenZhao** [comment]
> **breaking changes** to the table function API

What about introducing a version field to the UDF protocol? So that we can at least reject connecting to an old UDF server.

**Comment by wangrunji0408** [comment]
> > 1. If the function returns multiple columns, they will be put into the second column of the chunk as a struct.
>
> May I ask the reason for this reversion (#9635)? 👀

There are 2 reasons:
1. The return value of a table function is usually treated as a single column. In our code, the `return_type` of table functions is also a single `DataType`. Using a single column makes it consistent with scalar functions.
2. On the UDF server side, each table function call produces a struct value, which naturally forms a struct array. Using a single column in the protocol avoids unnecessary expansion to multiple columns.

**Comment by wangrunji0408** [comment]
> > **breaking changes** to the table function API
> 
> What about introducing a version field to the UDF protocol? So that we can at least reject connecting to an old UDF server.

Good idea!

File: src/udf/java/src/main/java/com/risingwave/functions/UdfClient.java
[the remark this answers is not in the document]
Comment (wangrunji0408) [reply src/udf/java/src/main/java/com/risingwave/functions/UdfClient.java:1]: For unit test...

**Comment by wangrunji0408** [comment]
> Good job. Except the comments above, one thing missing is the observability of udf server. To be exact, I want to provide two thing:
> 
> 1. In framework level, we should provide metrics of each function call.
> 2. We should allow user to construct their own metrics.

Is there something similar in Flink that I can refer to?

**Pull request merged**